consensus/consensus-types/src/pipelined_block.rs

```diff
@@ -120,6 +120,53 @@ pub struct PipelineInputRx {
     pub commit_proof_fut: TaskFuture<LedgerInfoWithSignatures>,
 }
 
+/// A window of blocks that are needed for execution with the execution pool, EXCLUDING the current block
+#[derive(Clone)]
+pub struct OrderedBlockWindow {
+    blocks: Vec<Weak<PipelinedBlock>>,
+}
+
+impl OrderedBlockWindow {
+    pub fn new(blocks: Vec<Arc<PipelinedBlock>>) -> Self {
+        Self {
+            blocks: blocks
+                .iter()
+                .map(Arc::downgrade)
+                .collect::<Vec<Weak<PipelinedBlock>>>(),
+        }
+    }
+
+    pub fn empty() -> Self {
+        Self { blocks: vec![] }
+    }
+
+    /// The blocks stored in `OrderedBlockWindow` use [`Weak`](Weak) pointers
+    ///
+    /// if the `PipelinedBlock` still exists
+    ///      `upgraded_block` will be `Some(PipelinedBlock)`, and included in `blocks`
+    /// else it will be `None`, and not included in `blocks`
+    pub fn blocks(&self) -> Vec<Block> {
```

Comment:
Should we have all-or-nothing semantics here? Wrapped in a `Result<>`?

Reply:
I think we're currently just unwrapping so will keep it as `panic!` and remove the `TODO` comment below